refactor: migrate `account-menu` to react

vardan-arm opened the pull request.

Commit:
feat: modify current tags to v4 style

Commit:
styles: fix class names

Commit:
feat: add autocomplete tags input and dropdown

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
chore(version-snjs): 2.3.2

Commit:
feat: omit active note tags from dropdown

Commit:
fix: add missing key

Commit:
fix: make dropdown height adjust to screen

Commit:
feat: add tag on dropdown option click

Commit:
feat: remove previous tag input

Commit:
fix: reload tag results after adding tag

Commit:
fix: keep tag title in dropdown in its original case

Commit:
fix: remove async from reloadActiveNoteTags

Commit:
refactor: extract Tag to its own component

Commit:
styles: make class names BEM and extract sn-dropdown-item style

Commit:
fix: add sync after adding tag

Commit:
feat: add new tag hint to dropdown

Commit:
fix: focus input after option click

Commit:
fix: reload tags after adding or removing them from context menu

Commit:
feat: create and add tag to note on hint click

Commit:
feat: create and add tag on click enter

Commit:
chore(version-snjs): 2.4.0

Commit:
feat: get sorted tags for note

vardan-arm committed:
refactor: migrate `account-menu` to react - initial UI

Commit:
fix: prevent form submission on dropdown option click

Commit:
fix: don't show hint if a tag already exists with that title

Commit:
feat: make hint fixed on bottom of dropdown

Commit:
feat: highlight tag on click

Commit:
feat: focus last tag when pressing backspace on input

Commit:
feat: remove tag on backspace press

Commit:
fix: make dropdown items full width

Commit:
fix: add streamItems event listener for tags

Commit:
fix: reload tag results after adding tag

Commit:
chore(version-snjs): 2.4.2

Commit:
chore(version): 3.7.1

Commit:
Merge branch 'release/3.7.1' into main

Commit:
refactor: remove legacy tags code from editor view

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
refactor: remove tags component from view

Commit:
fix: make hint visible and focus input after creating new tag

Commit:
styles: adjust input colors

Commit:
feat: add placeholder to tag input

Commit:
fix: only navigate to last tag on backspace if input is empty

Commit:
styles: fix tags padding

Commit:
styles: fix tags padding

Commit:
styles: constraint tags width

Commit:
styles: reduce tags max width

Commit:
feat: make tags container expandable

karolsojko committed:
feat: add configurable log level

karolsojko committed:
Merge branch 'hotfix/3.7.2'

karolsojko committed:
Merge branch 'hotfix/3.7.2' into develop

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- link the new React component to the app's store
- setup correct initial values
- small fixes

Commit:
feat: use panel width event instead of ResizeObserver

Commit:
refactor: refactor NoteTags for better readability

Commit:
feat: set tabindex depending on overflowed tags

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- implement different handlers, such as set/change/remove passcode, etc.
- setup correct initial values
- rename React component

Commit:
fix: fix tags container width

Commit:
feat: collapse tags on click outside

Commit:
styles: add animation for tags expanding/collapsing

Commit:
fix: revert to using fixed margin when setting container width

Commit:
fix: fix tags dropdown width and tags container height

Commit:
styles: fix hint tag styles

Commit:
feat: clicking on tag sets it as active

ainzzorl committed:
fix missing whitespace in permissions-modal (#568)

Commit:
feat: use opacity for overflowed tags animation

Commit:
fix: set tabIndex on dropdown buttons

Commit:
fix: re-add clearing tags ref on every render

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- handle login and register
- advanced options in login section
- focus/blur events with React
- cleanup password fields when required (setting the value to `undefined` for resetting the input field doesn't work well in React, thus set to empty string)

Commit:
fix: adjust layout for all zoom values and font sizes

Commit:
fix: fix typo and remove log

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
chore(deps): upgrade typescript

Commit:
fix: use ellipsis for tags text in dropdown

Commit:
fix: check if ref is present before setting overflow position

Commit:
feat: add delete tag button and refactor NoteTag to separate component

Commit:
fix: close dropdown after creating a tag

Commit:
fix: disregard modifiers for note selection if user hasn't triggered the action

Commit:
Merge branch 'develop' into bugfix/multiple-selection-incorrectly-enabled

Commit:
Merge pull request #569 from standardnotes/bugfix/multiple-selection-incorrectly-enabled

fix: disregard modifiers for note selection if user hasn't triggered the action

Commit:
fix: make sure input is visible when no tags are overflowed

Commit:
refactor: extract shared logic to active note state

Commit:
feat: context menu for deleting tags

Commit:
fix: reload layout when double clicking on panel resizer

Commit:
fix: reload tabIndex after tags expansion

Commit:
fix: fix focus on previous tag after tag deletion

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
feat: remove overflowed tags feature

Commit:
styles: reduce spacing when note has no tags

Commit:
styles: change tags icon color

Commit:
feat: show button to remove note on focus

Commit:
fix: onWidthEvent error

Commit:
fix: tags container margins

Commit:
feat: navigate tags with arrow keys

Commit:
styles: re-add accidentally removed icon color change

Commit:
fix: fix tags container max width

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

karolsojko committed:
fix: disable deprecations warnings

karolsojko committed:
Merge branch 'hotfix/3.7.3'

karolsojko committed:
Merge branch 'hotfix/3.7.3' into develop

karolsojko committed:
fix: silence deprecations on production

karolsojko committed:
Merge branch 'hotfix/3.7.4'

karolsojko committed:
Merge branch 'hotfix/3.7.4' into develop

Commit:
refactor: simplify state methods

Commit:
refactor: refactor autocomplete tag input in separate components and move shared logic to state

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- implement download backup
- implement import backup
- add `useCallback` to function that is used in `useEffect`, so that its reference isn't changed on every render

vardan-arm committed:
fix: use correct variable for password confirmation value

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

Commit:
refactor: rename state

Commit:
feat: add arrow key navigation for results dropdown

Commit:
fix: set max width to note tag text

Commit:
fix: spacing between title and tags

Commit:
fix: make tag selected on second click

Commit:
styles: set max width for tag hint

Commit:
fix: avoid event propagation when deleting a tag

Commit:
refactor: store refs in components

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
fix: add ellipsis overflow to tag text on note options

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
Merge branch 'feature/autocomplete-tags' of https://github.com/standardnotes/web into feature/autocomplete-tags

Commit:
fix: remove log

Commit:
fix: don't show dropdown if there are no results

Commit:
feat: focus input after selecting tag

Commit:
fix: add arrows navigation for tag hint

karolsojko commented on app/assets/javascripts/components/AccountMenuReact.tsx:
```diff
@@ -408,11 +412,74 @@ const AccountMenu = observer(({ application, appState, closeAccountMenu }: Props
+    if (!data) {
+      return;
+    }
+    if (data.version || data.auth_params || data.keyParams) {
```
could you turn this into something else? A cascade of if/else blocks with nested ifs can signify trouble in maintenance and debugging

vardan-arm replied:
Done

karolsojko commented on app/assets/javascripts/components/AccountMenuReact.tsx:
```diff
@@ -408,11 +412,74 @@ const AccountMenu = observer(({ application, appState, closeAccountMenu }: Props
+
+    setIsImportDataLoading(false);
+
+    if (!result) {
```
hmm maybe a switch block would be more in place instead of if..else if...else ?

vardan-arm replied:
There are different variables in each `if` condition but `switch` expects one expression inside it. Agree that this could be improved, made it more readable now.

karolsojko committed:
fix: JSON gem deprecation warnings

karolsojko committed:
Merge branch 'hotfix/3.7.5'

karolsojko committed:
Merge branch 'hotfix/3.7.5' into develop

vardan-arm committed:
refactor: reviewer's comment

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- implement "merge local data"
- show "error reporting dialog"
- fix TS errors
- cleanup TODOs

karolsojko committed:
feat: add alternative server for purposes of switching to API v1 (#571)

* feat: add alternative server for purposes of switching to API v1

* fix: naming of next version sync server host variable

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

karolsojko committed:
fix: missing next version server initialization

Commit:
styles: fix spacing between title and tag input

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

vardan-arm committed:
refactor: migrate `account-menu` to react - implement functionality

- remove all Angular code related to `account-menu`
- rename React component to AccountMenu so that many parts of old code remain unchanged
- code cleanup

Commit:
chore: run dev workflow on pull request event

Commit:
chore(version-snjs): 2.6.0

vardan-arm committed:
fix: fix typescript problem

Commit:
Merge branch 'develop' into account-menu-react

Commit:
Revert "chore: run dev workflow on pull request event"

This reverts commit 68239c1536b35ccf3db189b90b52aba54bf2034e.

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
feat: add parent chain tags when adding a tag

karolsojko committed:
Merge branch 'release/3.8.0'

vardan-arm committed:
refactor: reviewer's comments

- don't pass `closeAccountMenu` to AccountMenu, instead create it in Account menu state and use it
- add event observers as per original code
- move some variables into component's state (even if they don't need to have setters sometimes)

vardan-arm committed:
Merge remote-tracking branch 'origin/account-menu-react' into account-menu-react

vardan-arm committed:
refactor: reviewer's comments

- use state variables to decide whether email or passcode fields should be focused

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

Commit:
chore(version): 3.8.0

Commit:
Merge branch 'release/3.8.0' into main

Commit:
chore(version): 3.9.0-beta01

Commit:
Merge branch 'develop' into feature/autocomplete-tags

Commit:
Merge pull request #567 from standardnotes/feature/autocomplete-tags

feat: autocomplete tags

Commit:
feat: make delete backups checkbox off by default

Commit:
fix: make sessions modal scrollable

Commit:
styles: remove outline from tag input

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react

vardan-arm committed:
refactor: reviewer's comments

- cleanup event observers in `useEffect`
- other small changes

vardan-arm committed:
chore: run prettier

Commit:
chore(version-snjs): 2.6.1

Commit:
fix: if tag already exists it shouldn't be added to the list

Commit:
fix: tag rename to empty string should preserve its original title

Commit:
fix: add close on blur to tags dropdown

Commit:
styles: fix sessions modal height

Commit:
styles: change overflow to auto and add back dropdowns min width

Commit:
chore(version): 3.8.1

Commit:
fix: disallow creating empty tag

Commit:
chore(version-snjs): 2.6.2

Commit:
styles: remove min width from tags dropdown

vardan-arm committed:
Merge remote-tracking branch 'upstream/develop' into account-menu-react
